src/Features/Core/Portable/EmbeddedLanguages/Classification/EmbeddedLanguageClassifierContext.cs

```diff
@@ -8,52 +8,63 @@
+    /// </summary>
+    public SemanticModel SemanticModel { get; }
+
+    public CancellationToken CancellationToken { get; }
+
+    internal readonly ClassificationOptions Options;
+    internal readonly IVirtualCharService VirtualCharService;
+
+    internal EmbeddedLanguageClassificationContext(
+        SolutionServices solutionServices,
+        Project project,
+        SemanticModel semanticModel,
+        SyntaxToken syntaxToken,
+        TextSpan spanToClassify,
+        ClassificationOptions options,
+        IVirtualCharService virtualCharService,
+        SegmentedList<ClassifiedSpan> result,
+        CancellationToken cancellationToken)
+    {
+        SolutionServices = solutionServices;
+        Project = project;
+        SemanticModel = semanticModel;
+        SyntaxToken = syntaxToken;
+        _spanToClassify = spanToClassify;
+        Options = options;
+        VirtualCharService = virtualCharService;
+        _result = result;
+        CancellationToken = cancellationToken;
+    }
+
+    public void AddClassification(string classificationType, TextSpan span)
     {
-        internal readonly SolutionServices SolutionServices;
-
-        private readonly SegmentedList<ClassifiedSpan> _result;
-
-        public Project Project { get; }
-
-        /// <summary>
-        /// The string or character token to classify.
-        /// </summary>
-        public SyntaxToken SyntaxToken { get; }
-
-        /// <summary>
-        /// SemanticModel that <see cref="SyntaxToken"/> is contained in.
-        /// </summary>
-        public SemanticModel SemanticModel { get; }
-
-        public CancellationToken CancellationToken { get; }
-
-        internal readonly ClassificationOptions Options;
-        internal readonly IVirtualCharService VirtualCharService;
-
-        internal EmbeddedLanguageClassificationContext(
-            SolutionServices solutionServices,
-            Project project,
-            SemanticModel semanticModel,
-            SyntaxToken syntaxToken,
-            ClassificationOptions options,
-            IVirtualCharService virtualCharService,
-            SegmentedList<ClassifiedSpan> result,
-            CancellationToken cancellationToken)
-        {
-            SolutionServices = solutionServices;
-            Project = project;
-            SemanticModel = semanticModel;
-            SyntaxToken = syntaxToken;
-            Options = options;
-            VirtualCharService = virtualCharService;
-            _result = result;
-            CancellationToken = cancellationToken;
-        }
-
-        public void AddClassification(string classificationType, TextSpan span)
-            => _result.Add(new ClassifiedSpan(classificationType, span));
+        // Ignore characters that don't intersect with the requested span.  That avoids potentially adding lots of
+        // classifications for portions of a large string that are out of view.
+        if (span.IntersectsWith(_spanToClassify))
```

Comment:
note: in the case the user provided, while the string contains json, it's not a JsonTree.  All we ahve are the embedded classifications for escapes like `""` in the middle of the string.  But there are literally thousands of these classifications.  So this takes us down from many thousands of classifications to a few dozen.